### PR TITLE
SB - 123 [BE] Create table schemas

### PR DESCRIPTION
**Pull request by tonyxuan2021**
Screen Shot 2023-01-29 at 7 03 17 AM

- Created table models as per RDS 
- Connected to RDS (by making a query to the "/route-details/<string:route_id>" endpoint
- Enum data type was replaced by string because enum option (ie. cancelled) was already set up with RDS, tested with random string input and was prevented. Only predetermined enum options were allowed, therefore safe to use string data type.  
- Git stashed Aaron's testing code (ie. get all riders endpoint for RDS connection testing purpose)

### CODE REVIEW

File: models/Rider_route_requests.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+from decimal import Decimal
+
+
+@dataclass
+class Rider_route_requests(db.Model):
+    request_id: str = db.Column(UUID, primary_key=True, nullable=False)
+    user_id: str = db.Column(UUID, primary_key=True, nullable=False)
+    request_ts: int = db.Column(BigInteger)
+    pickup_addr: str = db.Column(db.String)
+    pickup_lat: Decimal = db.Column(db.Numeric(9, 6))
```
Comment (chunloy) [inline models/Rider_route_requests.py:14]: Question: do you think it's necessary to have high precision? I would think 2-3 decimals would be enough. To my understanding, precision refers to the total number of digits, and scale refers to the number decimals. 

Longitude coordinates range from +/-180, so maybe we could set it to `Numeric(5, 2)` or `Numeric(6, 3)`. 

Similarly, latitudes range from +/- 90, so maybe set it to `Numeric(5, 3)` or `Numeric(4, 2)`.

**Review by chunloy — CHANGES_REQUESTED** [review]
Good work on this Tony. I noticed there aren't any foreign keys in the models that need them. Can we add them as well? Have a look at the `db_v1.sql` file in the DB repo (specifically lines 143 - 159).

File: app.py [hunk 2]
```diff
@@ -50,16 +50,71 @@ def get_driver_routes():
...
+        waypoints_count=data["waypoints_count"],
+        polyline=data["polyline"],
+    )
+
+    # check for valid enum string
+    try:
+        db.session.add(route_data)
+        db.session.commit()
+    except Exception as e:
+        return e.args[0], 400
```
Comment (yhus) [inline app.py:91]: Use a more detailed error message here. e.g. "Failed to write routes data to database"

File: app.py [hunk 3]
```diff
@@ -50,16 +50,71 @@ def get_driver_routes():
...
+        or not data["created_at"]
+        or not data["predicted_start_secs"]
+        or not data["duration"]
+        or not data["status"]
+        or not data["waypoints_count"]
+        or not data["polyline"]
+    ):
+        return "One or more fields don't contain data", 401
+
+    route_data = Routes(
```
Comment (yhus) [inline app.py:76]: This part is business logic and should be factored out

File: app.py [hunk 4]
```diff
@@ -50,16 +50,71 @@ def get_driver_routes():
...
+    return "Successfully wrote route data to routes table", 200
+
+
+@app.route("/rider-details", methods=["GET"])
+def get_rider_details():
+    data = Riders.query.all()
+    return data
+
+
+@app.route("/rider-details", methods=["POST"])
```
Comment (yhus) [inline app.py:102]: Is this just for testing? If so, please leave a comment making that clear, or remove it if it's not needed

File: app.py [hunk 5]
```diff
@@ -88,3 +143,7 @@ def handle_driver_authentication(auth_token):
 
     except:
         return "Token is invalid", 401
+
+
+def get_route_details_by_route_id(routeId):
+    return Routes.query.filter_by(route_id=routeId).one_or_none()
```
Comment (yhus) [inline app.py:149]: Wrap with try catch, and proper exception handling error message.

Also, "business logic" should be moved out of app.py, into e.g. routes.py for this specific case

File: models/Routes.py [hunk 6]
```diff
@@ -1,11 +1,17 @@
+import enum
```
[the remark this answers is not in the document]
Comment (tonyxuan2021) [reply models/Routes.py:1]: removed.

File: models/Rider_route_requests.py (hunk 1 above)
Comment (tonyxuan2021) [reply models/Rider_route_requests.py:14]: Changed long to Numeric(6, 3), lat to Numeric(4, 2)

Comment (yhus) [reply models/Rider_route_requests.py:14]: http://wiki.gis.com/wiki/index.php/Decimal_degrees

For our use case I think we'd want at least 6  decimal places

Comment (tonyxuan2021) [reply models/Rider_route_requests.py:14]: Changed both lat and long to Numeric(8,6), indicating max 2 digit number with 6 decimal places.

File: app.py (hunk 2 above)
Comment (tonyxuan2021) [reply app.py:92]: Changed to "Failed to write routes data to database" under line 91.

File: app.py (hunk 3 above)
Comment (tonyxuan2021) [reply app.py:77]: Factored business logic (everything below line 76) to a helper function "handle_write_to_route_table".

File: app.py (hunk 4 above)
Comment (tonyxuan2021) [reply app.py:103]: Deleted "rider-details" both GET and POST endpoint since it was for testing use only.

**Review by yhus — APPROVED** [review]

**Review by StudentCZ — APPROVED** [review]
I saw you were missing a try/except for one of the endpoint functions but overall your table schemas looks good.

**Review by chunloy — APPROVED** [review]
👍

**Comment by tonyxuan2021** [comment]
Resolved merge conflict and addressed all comments.

**Pull request merged**